compiler/src/dotty/tools/dotc/util/Store.scala

```diff
@@ -0,0 +1,33 @@
+package dotty.tools.dotc.util
+
+object Store {
+
+  class Location[T](private[Store] val idx: Int) extends AnyVal
+
+  val empty: Store = new Store(Array())
+}
+
+class Store(private val elems: Array[AnyRef]) extends AnyVal {
+  import Store._
+
+  def newLocation[T](): (Location[T], Store) = {
+    val elems1 = new Array[AnyRef](elems.length + 1)
+    System.arraycopy(elems, 0, elems1, 0, elems.length)
+    (new Location(elems.length), new Store(elems1))
```

Comment:
For another PR: minor optimisations here to double the size of array and keep track of size?

Reply:
That would be complicated because arrays are purely functional. I.e you have to copy anyway on an update.